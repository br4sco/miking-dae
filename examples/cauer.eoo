
def serialize : Int -> (Node * Node -> Model) -> ((Node * Node) -> Model)
def serialize n f (n1, n2) =
    let ns =
      create (subi n 1) (\i. if eqi i (subi n 2) then n2 else gensym ())
    in
    let (eqns, n2) = foldl
      (\acc. \n1. let (eqns, n2) = acc in (eqns; f (n1, n2), n1))
      ([], n1) ns
    in
    eqns
end

main model
  def R = 1. end
  def L = 1. end
  def C = 1. end
  def k = 10 end
  node n1, n2, n3
  var uIn, uOut, t : Real
  model Section(n1, n2) = equation
    Inductor(L, n1, n2);
    Capacitor(C, n2, n3)
  end
equation
  initial t = 0.;
  t' = 1.;
  uIn = 1. / (1. + exp (0. - t));
  VoltageSource(uIn, n1, n2);
  serialize k Section (n1, n2);
  Capacitor(C, n3, n2);
  Resistor(R, n3, n2);
  VoltageSensor(uOut, n3, n2)
output
  (uIn, uOut)
end
