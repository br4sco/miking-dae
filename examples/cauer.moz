include GRAPHSIM
include GRAPHSIMEXPORT

def R = 1.
def L = 1.
def C = 1.

def Sections(n3: Node, acc: (Model, Node), n: Int) -> (Model, Node) = {
  if n < 1 then error "n is not a positive number"
  else
    def (sections, n1) = acc;
    def n2: Node;
    def acc = (Inductor(L, n1, n2); Capacitor(C, n2, n3); sections, n2);
    if n == 1 then acc
    else Sections(n3, acc, n - 1)
}

def CauerTopology(n: Int) = {
  def n1, n3: Node;
  def uIn, uOut: Voltage;
  def t: Signal;
  def (sections, n2) = Sections(n3, (VoltageSource(uIn, n1, n3), n1), n);
  t' = 1.;
  sections;
  Capacitor(C,n3,n2);
  Resistor(R,n3,n2);
  VoltageSensor(uOut, n3, n2);
  uIn = 1. / (1. + exp(-t))
}

def main =
  //export2ModelicaFile (elaborate (CauerTopology(50)))
  print (export2peadae (elaborate (CauerTopology(1000))) ++ "\n")
