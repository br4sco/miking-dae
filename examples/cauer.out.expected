-9.32910571867e-31
